rfc/rfc-40/rfc-40.md

```diff
@@ -0,0 +1,195 @@
+for processing. This is where the partition loader logic will reside. While listing the files for each Hudi partition
+the split manager will create one or more split per file. For non-partitioned table, the split mamager will simply
+return a single split for the entire table.
+
+During query execution, the Trino coordinator tracks all splits available for processing and the locations where tasks
+are running on workers and processing splits. As tasks finish processing and are producing more splits for downstream
+processing, the coordinator continues to schedule tasks until no splits remain for processing. Once all splits are
+processed on the workers, all data is available, and the coordinator can make the result available to the client.
+
+### Data source
+
+As mentioned in the query execution model, tasks in the source stage produce data in the form of pages. The Connector
+Data Source API returns pages when it is passed a split, and operators typically consume input pages, perform
+computation, and produce output pages. This is where we will implement `ConnectorPageSourceProvider` interface to create
+page source. We could have different page sources for different formats like parquet, orc and avro. For the data source,
+we plan to reuse the `ParquetPageSource` in the Hive connector. This has the advantage of using Trino's
+custom `ParquetReader` that can efficiently skip data sections by using statistics in file headers/footers. This is also
+where we will handle the column projections and build predicates for the parquet reader.
+
+```java
+public class HudiPageSourceProvider
+    implements ConnectorPageSourceProvider {
+  private final HdfsEnvironment hdfsEnvironment;
+  private final FileFormatDataSourceStats fileFormatDataSourceStats;
+  private final ParquetReaderOptions parquetReaderOptions;
+  private final DateTimeZone timeZone;
+}
+```
+In summary, Trino coordinator uses the metadata and split manager APIs to gather information about the table and partitions to
+generate a query plan and logical splits of the table contents. Each split is processed by a task in the Trino worker.
+Here, workers invoke the page source APIs as tasks produce data in the form of pages.
+Subsequently, native (parquet) reader read the block of pages while executing the query.
+
+### Improving split parallelism
+
+In order to improve split parallelism, we will push the fetch of partitions down to split source.
+Since each partition is independent of another, so we batch them up and each batch is processed by a thread.
+Additionally, the full list of partition names are fetched just once from the metastore and the subsequent
+construction of the partition key from the partition name can be done in parallel as well.
+
+### Improving listing
+In order to improve listing, we assume that the path exists,
+and so we bypass the `FileSystem#exists` check in `AbstractHoodieTableFileSystemView` while fetching latest base files.
```

Comment:
Guess MOR/Snapshot query is coming in a separate PR

Reply:
Yes. Tracked by HUDI-2740